datalogger/logger.go

```diff
@@ -95,3 +107,122 @@ func (logger *CSVDataLogger[T]) Close() bool {
 	logger.isOpen = false
 	return true
 }
+
+func CreateJSONDataLogger[T any](filename string) (DataLogger[T], error) {
+	fmt.Printf("Creating a JSON data logger: %v!\n", filename)
+	data := make([]T, 0)
+	destination, err := os.Create(filename)
+	if err != nil {
+		return &JSONDataLogger[T]{&sync.Mutex{}, 0, data, true, destination}, err
+	}
+
+	result := JSONDataLogger[T]{&sync.Mutex{}, 0, data, true, destination}
+	return &result, nil
+}
+
+func (logger *JSONDataLogger[T]) LogRecord(record T) {
+	logger.mut.Lock()
+	defer logger.mut.Unlock()
+	logger.recordCount += 1
+	logger.data = append(logger.data, record)
+}
+
+func PrettyPrint(out string, spaces int) string {
+	return fmt.Sprintf("%*s%s\n", spaces, "", out)
+}
+
+// Key Value Pair Print
+func KVPPrint(key string, value any) string {
+	return fmt.Sprintf("\"%s\": \"%v\"", key, value)
+}
+
+func (logger *JSONDataLogger[T]) Export() bool {
+	logger.mut.Lock()
+	defer logger.mut.Unlock()
+	if !logger.isOpen {
+		return false
+	}
+
+	// The rest of the code is written with the assumption that the indexing stays the same
+	visibleFields := reflect.VisibleFields(reflect.TypeOf((*T)(nil)).Elem())
+
+	// Meta Data
+	keys := make([]string, 0)
+	descriptions := make([]string, 0)
+	for _, v := range visibleFields {
+		// Keys
+		key, keysuccess := v.Tag.Lookup("json")
+		if !keysuccess {
+			key = fmt.Sprintf("None for %s", v.Name)
+		}
+		keys = append(keys, key)
+
+		// Descriptions
+		description, success := v.Tag.Lookup("Description")
+		if !success {
+			description = fmt.Sprintf("None for %s", v.Name)
+		}
+		descriptions = append(descriptions, description)
+	}
+
+	spaces := int(0)
+
+	// Open
+	logger.destination.Write([]byte(PrettyPrint("{", spaces)))
+	spaces += JSONSpaces
```

Comment:
`JSONSpaces` is for proper tabbing?

Reply:
Yes, set to 4 at the moment on line 26 :)